File: python_src/cartridge_sdk/ui/toast.py
"""Transient toast notifications."""

from __future__ import annotations

import time
from typing import TYPE_CHECKING

import pygame

from cartridge_sdk.screen import _get_font

if TYPE_CHECKING:
    from cartridge_sdk.theme import Theme


class Toast:
    """A single toast message with auto-dismiss."""

    def __init__(self, text: str, color: tuple = (220, 220, 230), duration: float = 2.5) -> None:
        self.text = text
        self.color = color
        self.created = time.monotonic()
        self.duration = duration

    @property
    def expired(self) -> bool:
        return time.monotonic() - self.created > self.duration

    @property
    def alpha(self) -> float:
        remaining = self.duration - (time.monotonic() - self.created)
        if remaining > 0.5:
            return 1.0
        return max(0.0, remaining / 0.5)


class ToastManager:
    """Manages a stack of toast notifications."""

    def __init__(self, max_toasts: int = 3) -> None:
        self._toasts: list[Toast] = []
        self._max = max_toasts

    def push(self, text: str, color: tuple = (220, 220, 230), duration: float = 2.5) -> None:
        self._toasts.append(Toast(text, color, duration))
        if len(self._toasts) > self._max:
            self._toasts.pop(0)

    def info(self, text: str, theme: Theme | None = None) -> None:
        color = theme.text if theme else (220, 220, 230)
        self.push(text, color)

    def success(self, text: str, theme: Theme | None = None) -> None:
        color = theme.text_success if theme else (100, 220, 100)
        self.push(text, color)

    def warn(self, text: str, theme: Theme | None = None) -> None:
        color = theme.text_warning if theme else (255, 200, 60)
        self.push(text, color)

    def error(self, text: str, theme: Theme | None = None) -> None:
        color = theme.text_error if theme else (255, 100, 100)
        self.push(text, color, duration=4.0)

    def draw(self, surface: pygame.Surface, theme: Theme, screen_w: int = 640, screen_h: int = 480) -> None:
        self._toasts = [t for t in self._toasts if not t.expired]
        if not self._toasts:
            return

        font = _get_font("mono", 14)
        lh = font.get_linesize()
        toast_h = lh + 12
        base_y = screen_h - 46  # above footer area

        for i, toast in enumerate(reversed(self._toasts)):
            y = base_y - (i + 1) * (toast_h + 4)
            alpha = toast.alpha

            # Background
            bg_surface = pygame.Surface((screen_w - 40, toast_h), pygame.SRCALPHA)
            bg_color = (*theme.bg_lighter, int(220 * alpha))
            bg_surface.fill(bg_color)
            surface.blit(bg_surface, (20, y))

            # Border
            border_surface = pygame.Surface((screen_w - 40, toast_h), pygame.SRCALPHA)
            border_color = (*toast.color, int(120 * alpha))
            pygame.draw.rect(border_surface, border_color, border_surface.get_rect(), 1)
            surface.blit(border_surface, (20, y))

            # Text
            text_surface = font.render(toast.text, True, toast.color)
            if alpha < 1.0:
                text_surface.set_alpha(int(255 * alpha))
            tx = 20 + (screen_w - 40 - text_surface.get_width()) // 2
            ty = y + (toast_h - lh) // 2
            surface.blit(text_surface, (tx, ty))

File: crates/cartridge-core/src/ui/toast.rs
//! Transient bottom-center toast notifications with auto-dismiss, colored left dot, and queue.

use sdl2::rect::Rect;

use crate::screen::{Screen, HEIGHT, WIDTH};

/// Toast severity level, determines the colored dot on the left.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ToastLevel {
    Info,
    Success,
    Warning,
    Error,
}

/// A single toast message with auto-dismiss timer.
pub struct Toast {
    pub text: String,
    pub level: ToastLevel,
    pub duration: f32,
    /// Time remaining in seconds.
    pub remaining: f32,
}

impl Toast {
    pub fn new(text: &str, level: ToastLevel, duration: f32) -> Self {
        Self {
            text: text.to_string(),
            level,
            duration,
            remaining: duration,
        }
    }

    /// Returns true if this toast has expired.
    pub fn expired(&self) -> bool {
        self.remaining <= 0.0
    }

    /// Alpha for fade-out in the last 0.5 seconds.
    pub fn alpha(&self) -> f32 {
        if self.remaining > 0.5 {
            1.0
        } else {
            (self.remaining / 0.5).max(0.0)
        }
    }
}

/// Manages a queue of toast notifications.
pub struct ToastManager {
    toasts: Vec<Toast>,
    max_toasts: usize,
}

impl ToastManager {
    pub fn new(max_toasts: usize) -> Self {
        Self {
            toasts: Vec::new(),
            max_toasts,
        }
    }

    /// Push a toast with explicit level and duration.
    pub fn push(&mut self, text: &str, level: ToastLevel, duration: f32) {
        self.toasts.push(Toast::new(text, level, duration));
        if self.toasts.len() > self.max_toasts {
            self.toasts.remove(0);
        }
    }

    /// Push an info toast.
    pub fn info(&mut self, text: &str) {
        self.push(text, ToastLevel::Info, 2.5);
    }

    /// Push a success toast.
    pub fn success(&mut self, text: &str) {
        self.push(text, ToastLevel::Success, 2.5);
    }

    /// Push a warning toast.
    pub fn warn(&mut self, text: &str) {
        self.push(text, ToastLevel::Warning, 2.5);
    }

    /// Push an error toast (longer duration).
    pub fn error(&mut self, text: &str) {
        self.push(text, ToastLevel::Error, 4.0);
    }

    /// Returns true if there are active toasts.
    pub fn has_toasts(&self) -> bool {
        !self.toasts.is_empty()
    }

    /// Advance timers by `dt` seconds. Call once per frame.
    pub fn update(&mut self, dt: f32) {
        for toast in &mut self.toasts {
            toast.remaining -= dt;
        }
        self.toasts.retain(|t| !t.expired());
    }

    /// Draw active toasts at the bottom center of the screen, above the footer area.
    ///
    /// Each toast is a rounded card with a colored dot on the left indicating the
    /// severity level, and centered text.
    pub fn draw(&self, screen: &mut Screen) {
        if self.toasts.is_empty() {
            return;
        }

        let theme = screen.theme;
        let font_size: u16 = 14;
        let lh = screen.get_line_height(font_size, false) as i32;
        let toast_h = lh + 12;
        let toast_w = (WIDTH as i32) - 40;
        let base_y = HEIGHT as i32 - 46; // above footer area

        for (i, toast) in self.toasts.iter().rev().enumerate() {
            let y = base_y - (i as i32 + 1) * (toast_h + 4);
            let _alpha = toast.alpha();

            // Background card
            let bg_rect = Rect::new(20, y, toast_w as u32, toast_h as u32);
            screen.draw_rect(bg_rect, Some(theme.bg_lighter), true, 4, None);

            // Border
            screen.draw_rect(bg_rect, Some(theme.card_border), false, 4, None);

            // Colored dot on the left based on level
            let dot_color = match toast.level {
                ToastLevel::Info => theme.text_accent,
                ToastLevel::Success => theme.positive,
                ToastLevel::Warning => theme.text_warning,
                ToastLevel::Error => theme.negative,
            };
            let dot_x = 20 + 10;
            let dot_y = y + toast_h / 2;
            screen.draw_circle(dot_x, dot_y, 4, dot_color);

            // Text -- centered in the remaining space after the dot
            let text_start_x = dot_x + 12;
            let text_area_w = toast_w - (text_start_x - 20) - 10;
            let tw = screen.get_text_width(&toast.text, font_size, false) as i32;
            let tx = text_start_x + (text_area_w - tw) / 2;
            let ty = y + (toast_h - lh) / 2;
            screen.draw_text(
                &toast.text,
                tx,
                ty,
                Some(theme.text),
                font_size,
                false,
                Some(text_area_w as u32),
            );
        }
    }
}

impl Default for ToastManager {
    fn default() -> Self {
        Self::new(3)
    }
}
